### PR TITLE
Remove imgui-go, implot, imnodes, ImGuiColorTextEdit from repo

### PR DESCRIPTION
**Pull request by ianling**
@AllenDang what are your thoughts on merging these changes upstream? 

I removed all of the hard copies of the dependencies in favor of importing them via a fork of imgui-go: https://github.com/ianling/imgui-go -- this fork is up-to-date with the upstream imgui-go repo as of yesterday, and also has all of the changes you've made in the giu repo.

If you are interested in changing to this style, we could merge this PR and I can transfer the fork ianling/imgui-go to your account. All instances of `ianling` would also need to be changed back to `AllenDang`

I think the main benefits of making this change are:

* reduces clutter by having a clear separation between the low-level CGo wrappers in imgui-go/implot/imnodes/ImGuiColorTextEdit and the clean high-level API provided by giu
* easier to maintain. If you need to make custom changes to imgui-go (or one of the other C++ libraries), you simply commit the change in the fork and import the new commit in giu. If the official imgui-go repo releases an update, or ImGui itself receives an update, it is very simple to update this fork. 
  * you can also update the imgui-go fork separately from giu, and continue building giu from the previous commit if there are breaking changes and you don't have time to update both repos at once.

### CODE REVIEW

**Comment by AllenDang** [comment]
The main reason why I used a hard copy of imgui-go is the efficiency to add or change wrappers from C++ part. I understand the benefit of decouple those parts, but a sync problem blocks me to do it.

For exmaple, if imgui-go is a standalone repo and once I've changed anything, idealy I could get the newest change by invoking `go get -u .../imgui-go@master`, but reality is this command doesn't always work, sometime I have to wait several hours to get the newest version. I don't know which part leads to it, maybe github's CDN or something else.

This issue will cost massive time.

**Comment by ianling** [comment]
I have also run into this issue, you can resolve it by using a commit hash instead of a branch name:    go get -u ...@12345abcdef6789abcdefThat will avoid the caching issue you see when using @masterOn Apr 14, 2021 19:50, Allen Dang ***@***.***> wrote:
The main reason why I used a hard copy of imgui-go is the efficiency to add or change wrappers from C++ part. I understand the benefit of decouple those parts, but a sync problem blocks me to do it.
For exmaple, if imgui-go is a standalone repo and once I've changed anything, idealy I could get the newest change by invoking go get -u ...***@***.***, but reality is this command doesn't always work, sometime I have to wait several hours to get the newest version. I don't know which part leads to it, maybe github's CDN or something else.
This issue will cost massive time.

**Comment by AllenDang** [comment]
Wow, this could work! Let's make this change. Decouple is always the better choice.

**Comment by ianling** [comment]
Cool, thank you! I will get the branch into a merge-able state and let you know when it is done.Do you want to just make a fork of ianling/imgui-go, or do you want me to transfer the repo? On Apr 14, 2021 20:05, Allen Dang ***@***.***> wrote:
Wow, this could work! Let's make this change. Decouple is always the better choice.

**Comment by AllenDang** [comment]
@ianling I prefer to transfer those repo to me. This will increase the efficiency when I tweak things from bottom to top.

**Comment by ianling** [comment]
@AllenDang -- I forgot to mention, we made a breaking API change in the giu fork, for HellSpawner. I can revert this or alter it if you would prefer not to keep it. The change is visible in the examples in the repo, here is part of the HelloWorld example (the change is the `.Build()` at the end of the loop function):

```
func loop() {
	g.SingleWindow("hello world").Layout(
		g.Label("Hello world from giu"),
		g.Line(
			g.Button("Click Me").OnClick(onClickMe),
			g.Button("I'm so cute").OnClick(onImSoCute),
			g.Button("Quit").OnClick(onQuit),
		),
	).Build()
}

func main() {
	wnd := g.NewMasterWindow("Hello world", 400, 200, g.MasterWindowFlagsNotResizable, nil)
	wnd.Run(loop)
}
```

This was to let us define a re-usable/template window, without making a full-blown widget:

```
helloWorldTemplate := g.SingleWindow("hello world").Layout(
		g.Label("Hello world from giu"),
		g.Line(
			g.Button("Click Me").OnClick(onClickMe),
			g.Button("I'm so cute").OnClick(onImSoCute),
			g.Button("Quit").OnClick(onQuit),
		),
	)

helloWorld1 := helloWorldTemplate
helloWorld2 := helloWorldTemplate

helloWorld1.Build()
helloWorld2.Build()
```

and also just as a performance improvement to prevent having to re-run `.Layout(<widget>, <widget>, <widget>, ...)` every time the window needs to be drawn. It lets us set the layout of a window once when we first create it, and then re-use that layout.

Additionally, we added some state tracking to WindowWidget to keep track of:
* does this WindowWidget have focus?
* the size of this window
* the position of this window

You can get these by doing (using the example from above):

```
hasFocus := helloWorld1.HasFocus()
size := helloWorld1.CurrentSize()
position := helloWorld1.CurrentPosition()

// there is also:
helloWorld1.BringToFront()
```

Maybe these changes do not fit with the "immediate mode" design, I'm not sure. I can remove it if you would prefer.

**Comment by AllenDang** [comment]
@ianling 
For the change to Layout, I don't quite understand whey we need to add a explicit Build there. We could just reuse the widgets by creating a variable outside loop so the widgets will not re-generate in each frame.

For the convinient methods for WindowWidget, I think they are helpful.

**Comment by ianling** [comment]
@AllenDang agreed, I have reverted that API change. All the examples are now the same as they were before, so there is no breaking API change. I think this is ready for review now.

**Comment by ianling** [comment]
I added some extra stuff to the multiwindow example demonstrating the methods I added to WindowWidget

**Comment by AllenDang** [comment]
@ianling Got it. This is s a complex PR, I may need sometime to check. :P

File: Context.go [hunk 1]
```diff
@@ -1,14 +1,14 @@
 package giu
 
 import (
-	"sync"
+    platform2 "github.com/AllenDang/giu/platform"
```
Comment (AllenDang) [inline Context.go:4]: platform2 is a bad name. Suggest to change to imguiPlatform for better understanding, so to renderer.

**Review by AllenDang — APPROVED** [review]

**Comment by AllenDang** [comment]
@ianling I just found out that imgui-go is not using the newest version from my repo. I'm using GLFW v3.3 with gl v3.3. And I have to check what else is missing now...

**Comment by AllenDang** [comment]
@ianling Oh my.... we've implement the platform and renderer in imgui-go repo, why the reference path in giu still is github.com/AllenDang/giu/platform?

**Comment by AllenDang** [comment]
@ianling I have to revert this PR because it uses a very old version of imgui-go. And the platform and renderer is still couple inside giu.

**Comment by ianling** [comment]
Will look into this, sorry, must have missed these.

**Comment by gucio321** [comment]
> The main reason why I used a hard copy of imgui-go is the efficiency to add or change wrappers from C++ part. I understand the benefit of decouple those parts, but a sync problem blocks me to do it.
> 
> For exmaple, if imgui-go is a standalone repo and once I've changed anything, idealy I could get the newest change by invoking go get -u .../imgui-go@master, but reality is this command doesn't always work, sometime I have to wait several hours to get the newest version. I don't know which part leads to it, maybe github's CDN or something else.
> 
> This issue will cost massive time.

@AllenDang I just found an excelent solution: https://blog.gopheracademy.com/advent-2015/vendor-folder/
so it looks like if you do `go mod vendor` you'll get `vendor` direcory with all your dependencies (e.g. imgui-go) which you can edit and test them with giu on-time.

**Pull request merged**